app/src/main/java/io/github/droidkaigi/confsched2017/view/customview/ParticlesAnimationView.java

```diff
@@ -100,68 +122,43 @@ public Shader resize(int width, int height) {
     private List<Particle> createParticles(int count) {
         List<Particle> particles = new ArrayList<>();
         for (int i = 0; i < count; i++) {
-            particles.add(new Particle(getWidth(), getHeight()));
+            particles.add(new Particle(getWidth(), getHeight(), this));
         }
         return particles;
     }
 
-    private List<Line> createLines(List<Particle> particles) {
-        final List<Line> lines = new ArrayList<>();
-        Stream.of(particles)
-                .forEach(particle -> Stream.of(particles)
-                        .filter(targetParticle ->
-                                particle.isShouldLinked(LINK_HEXAGON_DISTANCE, targetParticle)
-                        )
-                        .map(linkParticle -> new Line(new float[]{
-                                particle.center.x,
-                                particle.center.y,
-                                linkParticle.center.x,
-                                linkParticle.center.y})
-                        )
-                        .forEach(line -> {
-                            if (!lines.contains(line)) lines.add(line);
-                        })
-                );
-        return lines;
-    }
-
-    private class Line {
+    /**
+     * A Pair of 'Particles' whose centers can be 'linked to each other ... Called 'Line'.
+     */
+    private static class Line extends Pair<Particle, Particle> {
 
         private static final int MAX_ALPHA = 172;
-        private float[] point;
 
-        public Line(float[] point) {
-            this.point = point;
+        /**
+         * Constructor for a Pair.
+         *
+         * @param first  the first object in the Pair
+         * @param second the second object in the pair
+         */
+        Line(Particle first, Particle second) {
+            super(first, second);
         }
 
-        public void draw(Canvas canvas, Paint paint) {
-            final double distance = Math.floor(Math.sqrt(
-                    (point[2] - point[0]) * (point[2] - point[0])
-                            + (point[3] - point[1]) * (point[3] - point[1])
-            ));
+        void draw(Canvas canvas, Paint paint) {
+            if (!first.shouldBeLinked(LINK_HEXAGON_DISTANCE, second)) {
```

Comment:
@roana0229 not really get your full point, but do you mean to calling ```shouldBeLinked``` from ```ParticlesAnimationView#onDraw``` instead of here? It will end up removing the need of using "Line" object (a Line is actually just a bridge between 2 particles :D).

Yesterday I thought about something like:

```java
    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        for (int i = 0, size = particles.size(); i < size; i++) {
            particles.get(i).draw(canvas, paint);
        }

        // for (int i = 0, size = lines.size(); i < size; i++) {
        //    lines.get(i).draw(canvas, paint);
        //}

        for (int i = 0; i < particles.size(); i++) {
            for (int j = 0; j < particles.size(); j++) {
                if (particles.get(i).shouldBeLinked(LINK_HEXAGON_DISTANCE, particles.get(j))) {
                    // TODO draw, use the code from Line
                }
            }
        }
    }
```

I think what you mean is the following:

```java
    // ParticlesAnimationView
    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        for (int i = 0, size = particles.size(); i < size; i++) {
            particles.get(i).draw(canvas, paint);
        }

         for (int i = 0, size = lines.size(); i < size; i++) {
             if (lines.get(i).shouldDraw()) {
                 lines.get(i).draw(canvas, paint);
             }
        }
    }

    // ParticlesAnimationView$Line
    ///// shouldDraw simply returns the comparison
    boolean shouldDraw() {
        return first.shouldBeLinked(LINK_HEXAGON_DISTANCE, second);
    }
```

Either of above gain any huge impact (but the later looks clean and better, in fact :D). For now I would like to keep current version (this PR is pretty huge now) and maybe another PR from you if you really want to :p. So it is up to you ^^.